### ISSUE TITLE
ZIO App does not exit on OutOfMemoryError

### ISSUE THREAD

**Issue by Marc-Ducret** [zio/zio#4452]
The following program:
```scala
import zio._
import zio.duration._

object ZioMain extends App {
  override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] =
    ZIO
      .effectTotal {
        println("pouf")
      }
      .delay(1.second)
      .forever
      .fork *>
      ZIO.effectTotal {
        val n = 1024 * 1024 * 1024
        Array.fill(n)(0)
      }.exitCode
}
```
with the following dependencies versions:
```scala
object Versions {
  val catsEffect = "2.2.0"
  val zio = "1.0.3"
  val ziocats = "2.2.0.1"
}
```
produces the following output:
```
java.lang.OutOfMemoryError: Java heap space
	at scala.reflect.ManifestFactory$IntManifest.newArray(Manifest.scala:221)
	at scala.reflect.ManifestFactory$IntManifest.newArray(Manifest.scala:219)
	at scala.Array$.fill(Array.scala:347)
	at ZioMain$.$anonfun$run$3(ZioMain.scala:15)
	at ZioMain$$$Lambda$101/1961748891.apply(Unknown Source)
	at zio.internal.FiberContext.evaluateNow(FiberContext.scala:398)
	at zio.internal.FiberContext.$anonfun$fork$17(FiberContext.scala:772)
	at zio.internal.FiberContext$$Lambda$98/558187323.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
pouf
pouf
pouf
pouf
pouf
pouf
pouf
pouf
pouf
pouf
```
(and keeps going)

I would expect all fibers to be interrupted when such a fatal error occurs.

**Comment by adamgfraser** [comment]
I'm not sure this behavior is incorrect. When a fatal error occurs the system is in an undefined state and normally the best we can do is fail fast versus trying to do more work attempting to shut down gracefully that will itself use memory, which we just said we don't have any of.

**Comment by Marc-Ducret** [comment]
Here there is no shutdown at all, the program and in particular the fiber that prints "pouf" keeps running forever.

**Comment by Marc-Ducret** [comment]
If you confirm that it is intended behavior for a ZIO program to **not** exit when a fiber encounters a OutOfMemoryError, then I guess this issue can be closed.

Intended or not, if anyone else want a different behavior, it can be achieved easily doing by overriding the platform field of your `zio.App`:
```scala
  override val platform: Platform = {
    Platform.default
      .withReportFatal { error =>
        try {
          java.lang.Runtime.getRuntime.halt(-1)
          throw error
        } catch {
          case _: Throwable => throw error
        }
      }
  }
```

**Comment by adamgfraser** [comment]
Closed by #4812.
